refactor!: introducing `#[external(...)]` macro

benesjan opened the pull request:
Closes https://linear.app/aztec-labs/issue/F-19/rename-privateorpublicorutility-to-externalprivateorpublicorutility

Having `#[private]` and `#[public]` macros was misleading because in Solidity these define function visibility while in Aztec's case they define the environment in which the functions are executed. For this reason I introduce `#[external(...)]` macro while the argument of the attribute defines the execution environment: `private`, `public` or `utility`.

## Notes for reviewer
- I did a refactor in a [PR up the stack](https://github.com/AztecProtocol/aztec-packages/pull/17840) which allowed me to drop a lot of the complexity introduced in this PR. For this reason I encourage you to check that one first to have context on what in this PR is a problem and what is not. It might make sense to just merge these 2 PRs into 1 but I think that would make it all harder to review.
- This PR touches a lot of files but all the juicy bits are in the `noir-projects/aztec-nr/aztec/src/macros/` directory so I encourage you to just check that first. The `noir-contracts` dir can be completely ignored as there I just update the imports and the macros.
- The new naming is now in conflict with the `#[internal]` macro as having both external and internal applied to a function is valid:

<img width="874" height="182" alt="image" src="https://github.com/user-attachments/assets/4582a51a-225c-4a4c-9b9a-d2c2d413d927" />

This issue will be tackled in the near future.

benesjan commented:
* **#17963** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17963?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17997** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17997?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17958** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17958?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17884** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17884?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17840** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17840?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17827** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17827?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17827?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `next`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

benesjan commented on noir-projects/aztec-nr/aztec/src/macros/functions/mod.nr:
```diff
@@ -35,10 +44,13 @@ use auth_registry::AUTHORIZE_ONCE_REGISTRY;
     // Marker attribute - see the comment above
 
-    if !is_fn_private(f) & !is_fn_public(f) {
+    if !is_fn_external(f) {
```
This is the main sad thing about this PR. Now I am unable to evaluate here if the function is not utility because I don't have the guarantee if the external macro function has already been run. So I only perform this check here and then in utility macro I check if these attributes are not applied.

nventuro replied:
Yea this is quite sad. What might be a direction long term? Would we keep doing 
```noir
#[initializer]
#[external("private")
fn foo
```

or might we move to e.g.
```noir
#[external("private", "initializer")]
fn bar

#[external("private", "no_init_check", "view")]
fn qux
```

In that scenario we'd not need to worry about macro order or not having the full picture.

benesjan replied:
I would go with the second approach as it will make the macros much clearer.

The only bad thing about it is that invalid attribute args currently don't play well with LSP:

<img width="2040" height="1712" alt="image" src="https://github.com/user-attachments/assets/4ba4fee3-7ab2-4f8b-8390-3f7ff543811c" />

But I assume that Noir team could improve that and give us a nice error message at which point all seems fine.

benesjan commented on noir-projects/aztec-nr/aztec/src/macros/functions/mod.nr:
```diff
@@ -98,37 +119,62 @@ pub comptime fn authorize_once(
+/// Private functions are executed client-side and preserve privacy.
+comptime fn private(f: FunctionDefinition) -> Quoted {
+    // We need to add this attribute to be able to identify if the function is external private in TS.
+    f.add_attribute("private");
```
Another sad thing. Needed to do this as otherwise only the "external" attribute travels to the JSON artifact and in TS we need to know the external type.

nventuro replied:
This bit seems fine, not concerned.

benesjan commented on noir-projects/aztec-nr/aztec/src/macros/utils.nr:
```diff
@@ -1,26 +1,43 @@
 use dep::protocol_types::meta::derive_serialize;
 use std::meta::unquote;
 
 pub(crate) comptime fn get_fn_visibility(f: FunctionDefinition) -> Quoted {
```
The naming here is broken and when thinking about how to rename this I realized that the best thing to do is to just drop this so I did that in a [PR up the stack](https://github.com/AztecProtocol/aztec-packages/pull/17840).

I am aware the body of this function could be cleaned up but makes no sense to bother with it.

nventuro commented on docs/docs/developers/migration_notes.md:
```diff
@@ -171,6 +171,42 @@ The following commands were dropped from the `aztec` command:
+### Replacing #[private], #[public], #[utility] with #[external(...)] macro
+
+The original naming was not great in that it did not sufficiently communicate what the given macro did.
+We decided to rename `#[private]` as `#[external("private")]`, `#[public]` as `#[external("public")]`, and `#[utility]` as `#[external("utility")]` to better communicate that these functions are externally callable and to specify their execution context.
```
```suggestion
We decided to rename `#[private]` as `#[external("private")]`, `#[public]` as `#[external("public")]`, and `#[utility]` as `#[external("utility")]` to better communicate that these functions are externally callable and to specify their execution context. In this sense, `external` now means the exact same thing as in Solidity, i.e. a function that can be called from other contracts, and that can only be invoked via a contract call (i.e. the `CALL` opcode in the EVM, and a kernel call/AVM `CALL` opcode in Aztec).
```

nventuro commented on noir-projects/aztec-nr/aztec/src/macros/functions/mod.nr:
```diff
@@ -98,37 +118,56 @@ pub comptime fn authorize_once(
+    } else {
         panic(
-            f"A function marked as #[private] cannot also be #[public] or #[utility] - {name} is more than one of these",
+            f"Invalid external function type - {f_type} is not a valid function type. Choose one of private, public, or utility",
```
A better error message might be:

```
Function 'foo' is marked as #[external("potato")], but 'potato' is not a valid external function type. 

External functions must be one of 'private', 'public' or 'utility'.
```

So that it's much more immediately obvious to the reader which fn we're talking about (foo), and what is it they did that was wrong (#[extenrnal] was ok, the potato part was wrong).

benesjan replied:
That's indeed better. Addressed in https://github.com/AztecProtocol/aztec-packages/pull/17827/commits/24854f2514f980d63cba46a2090b2bf4ebadf0a7

Thanks

nventuro commented on noir-projects/aztec-nr/aztec/src/macros/functions/mod.nr:
```diff
@@ -98,37 +118,56 @@ pub comptime fn authorize_once(
+
+/// Private functions are executed client-side and preserve privacy.
+comptime fn private(f: FunctionDefinition) -> Quoted {
+    // We need to add this attribute to be able to identify if the function is external private in TS.
```
Instead of 'in TS' I'd instead say something like 'when constructing the contract artifact' or whatever.

benesjan replied:
Addressed in https://github.com/AztecProtocol/aztec-packages/pull/17827/commits/d320b15503358e7dfad067483f2203c0dfa32837

nventuro commented on noir-projects/noir-contracts-comp-failures/contracts/invalid_external_function_type/src/main.nr:
```diff
@@ -0,0 +1,9 @@
+pub contract InvalidExternalFunctionType {
+    use aztec::macros::functions::external;
+
+    #[external("invalid")]
+    fn invalid_external_function_type() {}
```
Good boy

nventuro reviewed (commented):
Very nice, thanks a lot! Leaving the review as a comment as I want to a) confirm with product we'll not accidentally release this until we're ready, and b) discuss a bit how we'll deal with other fn attributes like view, initializer, etc.

benesjan commented:
Seems to have been accidentally closed by Jake so reopening

benesjan committed:
refactor!: introducing `#[external(...)]` macro

Closes https://linear.app/aztec-labs/issue/F-19/rename-privateorpublicorutility-to-externalprivateorpublicorutility

Having `#[private]` and `#[public]` macros was misleading because in Solidity these define function visibility while in Aztec's case they define the environment in which the functions are executed. For this reason I introduce `#[external(...)]` macro while the argument of the attribute defines the execution environment: `private`, `public` or `utility`.

## Notes for reviewer
- I did a refactor in a [PR up the stack](https://github.com/AztecProtocol/aztec-packages/pull/17840) which allowed me to drop a lot of the complexity introduced in this PR. For this reason I encourage you to check that one first to have context on what in this PR is a problem and what is not. It might make sense to just merge these 2 PRs into 1 but I think that would make it all harder to review.
- This PR touches a lot of files but all the juicy bits are in the `noir-projects/aztec-nr/aztec/src/macros/` directory so I encourage you to just check that first. The `noir-contracts` dir can be completely ignored as there I just update the imports and the macros.
- The new naming is now in conflict with the `#[internal]` macro as having both external and internal applied to a function is valid:

<img width="874" height="182" alt="image" src="https://github.com/user-attachments/assets/4582a51a-225c-4a4c-9b9a-d2c2d413d927" />

This issue will be tackled in the near future.

Co-authored-by: Jan Beneš <janbenes1234@gmail.com>
